File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.kt
class Solution {
    fun longestValidParentheses(s: String): Int {
        var ans = 0
        var left = 0
        var right = 0
        for (ch in s) {
            if (ch == '(') left++ else right++
            if (left == right) ans = maxOf(ans, 2 * right)
            else if (right > left) { left = 0; right = 0 }
        }
        left = 0; right = 0
        for (i in s.indices.reversed()) {
            if (s[i] == ')') right++ else left++
            if (left == right) ans = maxOf(ans, 2 * left)
            else if (left > right) { left = 0; right = 0 }
        }
        return ans
    }
}

File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.swift
class Solution {
    func longestValidParentheses(_ s: String) -> Int {
        var ans = 0, left = 0, right = 0
        for ch in s {
            if ch == "(" { left += 1 } else { right += 1 }
            if left == right { ans = max(ans, 2 * right) }
            else if right > left { left = 0; right = 0 }
        }
        left = 0; right = 0
        for ch in s.reversed() {
            if ch == ")" { right += 1 } else { left += 1 }
            if left == right { ans = max(ans, 2 * left) }
            else if left > right { left = 0; right = 0 }
        }
        return ans
    }
}